http/src/main/scala/org/http4s/blaze/http/http20/Http2ServerHubStage.scala

```diff
@@ -0,0 +1,517 @@
+package org.http4s.blaze.http.http20
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets.UTF_8
+import java.util
+
+import org.http4s.blaze.pipeline.Command.OutboundCommand
+import org.http4s.blaze.pipeline.stages.addons.WriteSerializer
+import org.http4s.blaze.pipeline.{Command => Cmd, LeafBuilder}
+import org.http4s.blaze.pipeline.stages.HubStage
+import org.http4s.blaze.util.BufferTools
+import org.http4s.blaze.util.Execution.trampoline
+
+import Settings.{ DefaultSettings => Default, Setting }
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{Promise, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+// TODO: is it best to stick with extending the WriteSerializer?
+final class Http2ServerHubStage[HType](headerDecoder: HeaderDecoder[HType],
+                                       headerEncoder: HeaderEncoder[HType],
+                                        node_builder: () => LeafBuilder[NodeMsg.Http2Msg[HType]],
+                                             timeout: Duration,
+                                   maxInboundStreams: Int,
+                                       inboundWindow: Int = Default.INITIAL_WINDOW_SIZE)
+  extends HubStage[ByteBuffer] with WriteSerializer[ByteBuffer]
+{ hub =>
+
+  import Http2Exception._ // used by pretty much all the error handling
+  import bits._
+  
+  require(maxInboundStreams > 0, "Invalid max streams")
+
+  override def name: String = "Http2ServerHubStage"
+
+  private implicit def ec = trampoline
+
+  type Http2Msg = NodeMsg.Http2Msg[HType]
+
+  override type Key = Int
+  override type Out = Http2Msg
+  override protected type Attachment = FlowControl.NodeState
+
+  // Using synchronization and this is the lock
+  private val lock = new AnyRef
```

Comment:
Events from streams modify things like flow control windows, the header compressor state, making and closing streams etc. Some events depend on many of the different parts of state. The only way to maintain all of it atomically would be to have a super state type which would be swapped out atomically, or technically an actor is based on atomic operations.